numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -1592,3 +1593,302 @@ array_datetime_as_string(PyObject *NPY_UNUSED(self), PyObject *args,
+    newfmt[outcount] = 0;
+    if (dts.year == NPY_DATETIME_NAT) {
+        buff[0] = 'N';
+        buff[1] = 'a';
+        buff[2] = 'T';
+        buff[3] = '\0';
+    }
+    else {
+        /* Convert that into a string */
+        /* Gregorian calender repeats every 400 y, so
+        put within nearby epoch; note year and subsecond formatting
+        handled above, so this just formats everything else.*/
+        div_t output2 = div(dts.year, 400);
+        dts.year = output2.rem + 2000;
+
+        if (get_formatted_from_datetimestruct(newfmt, &dts, buff) < 0){
+            return -1;
+        }
+    }
+    return 0;
+}
+
+
+/*
+ * This is the Python-exposed datetime_strftime function.
+ */
+NPY_NO_EXPORT PyObject *
+array_datetime_strftime(PyObject *NPY_UNUSED(self), PyObject *args,
+                                PyObject *kwds)
+{
+    PyObject *arr_in = NULL, *fmt_in = NULL, *timezone_obj = NULL;
+
+    PyArray_DatetimeMetaData *meta;
+    int strsize;
+    PyArrayObject *ret = NULL;
+
+    NpyIter *iter = NULL;
+    PyArrayObject *op[2] = {NULL, NULL};
+    PyArray_Descr *op_dtypes[2] = {NULL, NULL};
+    npy_uint32 flags, op_flags[2];
+    char *fmt;
+
+    static char *kwlist[] = {"arr", "fmt", "timezone", NULL};
+
+    if(!PyArg_ParseTupleAndKeywords(args, kwds,
+                                "OO|O:datetime_strftime", kwlist,
+                                &arr_in,
+                                &fmt_in,
+                                &timezone_obj)) {
+        return NULL;
+    }
+
+    /* Claim a reference to timezone for later */
+    Py_XINCREF(timezone_obj);
+
+    op[0] = (PyArrayObject *)PyArray_FROM_O(arr_in);
+    if (op[0] == NULL) {
+        goto fail;
+    }
+    if (PyArray_DESCR(op[0])->type_num != NPY_DATETIME) {
+        PyErr_SetString(PyExc_TypeError,
+                    "input must have type NumPy datetime");
+        goto fail;
+    }
+
+    /* Get the datetime metadata */
+    meta = get_datetime_metadata_from_dtype(PyArray_DESCR(op[0]));
+    if (meta == NULL) {
+        goto fail;
+    }
+
+    if (fmt_in != NULL && fmt_in != Py_None) {
+        PyObject *strobj;
+        char *str = NULL;
+        Py_ssize_t len = 0;
+
+        if (PyUnicode_Check(fmt_in)) {
+            strobj = PyUnicode_AsASCIIString(fmt_in);
+            if (strobj == NULL) {
+                goto fail;
+            }
+        }
+        else {
+            strobj = fmt_in;
+            Py_INCREF(strobj);
+        }
+
+        if (PyBytes_AsStringAndSize(strobj, &str, &len) < 0) {
+            Py_DECREF(strobj);
+            goto fail;
+        }
+        fmt = str;
+    }
+    else {
+      PyErr_SetString(PyExc_TypeError, "fmt argument must be a string");
+      goto fail;
+    }
+
+    /* Get a string size long enough for any datetimes we're given
+    note we can't really know this ahead of time unless we format
+    all the strings because the year field can be arbitrarily
+    long */
+    strsize = strlen(fmt) + 20;
```

Comment:
Thats true; if the user asks for repeated %Y or repeated %f this can overflow.  I'll put in a catch for that somehow...

Reply:
If you cannot give a reasonable upper bound on the strftime (because it is too tricky or just to large). I think we should consider just making the output an object array (maybe documenting that it may be converted to a compatible variable sized string array in some future).

Reply:
Well I think the proposed length is reasonable.  It’s hard to think of a good reason to format a year multiple times or the subseconds multiple times.  Similarly you can’t get subseconds to many decimal places before you can’t represent years to large lengths.  So the overflow would be for people formatting %Y and/or %18f many times.  But it’s true that should error and not segfault

Reply:
Again, I think the cleanest way to avoid this problem would be to use an object array of Python strings instead of a Unicode array.  I *think* this should be straightforward, and an actual simplification of the code, but I haven't tried it yet.

Reply:
At the point we start to use objects, I'd start to argue we should drop the vectorization entirely, and let the caller do so - in which case, `datetime_strftime(d, fmt)` would be renamed to `datetime64.__format__(d, fmt)`

Reply:
I propose leaving it as is, and putting a catch in somewhere to make sure the returned string  never gets too large.